modules/electrophysiology_browser/php/models/electrophysioevents.class.inc

```diff
@@ -210,51 +286,95 @@ class ElectrophysioEvents
             ['PFID' => $this->_physioFileID]
         );
 
-        if (count($tsv) > 0) {
-            $tsvPath = $dataDir . $tsv[0]['filePath'];
-            // Update files if files updated before database updated
-            if ($tsv[0]['lastWritten'] <= $tsv[0]['lastUpdate']) {
-                // events.tsv
-                $tsvFile = fopen($tsvPath, 'w'); // Will override all file content
+        $projectID = intval(
+            $db->pselectOne(
+                'SELECT ProjectID FROM session AS s WHERE s.ID = (
+                SELECT SessionID FROM physiological_file
+                WHERE PhysiologicalFileID=:PFID
+            )',
+                ['PFID' => $this->_physioFileID]
+            )
+        );
 
-                $extraColumns = $db->pselect(
-                    "SELECT *
+        $json = $db->pselect(
+            "SELECT
+        EventFileID   AS id,
+        FilePath      AS filePath,
+        LastUpdate    AS lastUpdate,
+        LastWritten   AS lastWritten
+      FROM physiological_event_file
+      WHERE (
+          PhysiologicalFileID=:PFID OR ProjectID=:PID
+      )
+      AND FileType='json'",
+            [
+                'PFID' => $this->_physioFileID,
+                'PID'  => $projectID
+            ]
+        );
+
+        $tsvPath  = count($tsv) > 0 ? $dataDir . $tsv[0]['filePath'] : '';
+        $jsonPath = count($json) > 0 ? $dataDir . $json[0]['filePath'] : '';
```

Comment:
The file I/O here is not done safely. I haven't determined where the filePath comes from, but if it comes from the user (or ever might) this is very insecure. A user can craft a path with an arbitrary number of `../../` to overwrite any file on the filesystem that the web server has write access to. There is also no check around the `fopen` call below for the `''` case.

Luckily the fix is fairly simple. Use `\Utility::joinPath($dataDir, \Utility::resolvePath(filepath))` in all the places that calculate a path. This will resolve all the relative components of filePath before joining it to dataDir so that the path calculated can't escape $dataDir (and add an extra check for `''` or `$dataDir` in the if statements around the fopen calls.)

Reply:
Resolved by fcb9fe5efd802426b34de279723d4061543e196d

Using Utility functions